text/0069-api-v2.md

```diff
@@ -0,0 +1,106 @@
+
+## Document-level design
+
+This RFC introduces a new key encode to RawKV and TxnKV, and a new value encode to RawKV, which will allow the RawKV to be used along with TxnKV and also allow TiKV to flexibly add meta, e.g. the TTL, to a RawKV value.
+
+The `API V2` is enabled by a switch on PD. Since it changes the storage encode, it will be not compatible to switch between `API V1` and `API V2` while there are non-TiDB data in TiKV. TiDB data is specially treated in order to not be affected by the change.
+
+### Upgrade
+
+Upgrade from `API V1` to `API V2` is a simple process:
+
+1. Update TiKV, TiDB, and PD to the version that supports `API V2`.
+2. Ensure that no non-TiDB data is in TiKV. Detele if any. (Guaranteed by user)
+3. Ensure that no non-TiDB client is online. (Guaranteed by user)
+4. Use `pd-ctl` to enable `API V2`.
+5. Restart all TiKV. (Guaranteed by user)
+
+### Downgrade
+
+Downgrade from `API V2` to `API V1` is also simple:
+
+1. Ensure that no non-TiDB TiKV client is online.
+2. Ensure that no non-TiDB data is in TiKV. Detele if any. (Guaranteed by user)
+3. Use `pd-ctl` to disable `API V2`.
+4. Restart all TiKV. (Guaranteed by user)
+
+### Client
+
+TiKV clients will have a new configuration choosing `API V1` or `API V2` mode on connecting to TiKV.
+
+TiKV will reject the request if the client mode is mismatched to the TiKV configuration. There is one exception: With TiKV in `API V2` mode, you can use `TxnClient` in `API V1` mode to access the TiDB data (key start with `m` or `t`) which is invisible to the client in `API V2` mode.
+
+Note that in `API V2` mode, you can't specify CF in `RawClient`.
+
+Using TxnKV along with RawKV is safe in `API V2` mode.
+
+## Reference-level design
+
+### RawKV Key Encode
+
+Once the `API V2` is enabled, the key will be either start with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `x\0`: TxnKV key. The `\0` is a placeholder for the potential [Keyspace](https://github.com/tikv/rfcs/pull/39).
+3. `r\0`: RawKV key.
+
+Note that in TxnKV, the key will be encoded by `Memory Comparable Encoding`. But since the `Memory Comparable Encoding` will not change the starting bytes but only add paddings, there won't be any overlap between RawKV and TxnKV.
+
+### RawKV Value Encode
+
+If the key is started with `r`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### PD
+
+Add a new configuration `enable_api_v2` which can be configured by `pd-ctl`. When switching on or off, PD will synchronize the configuration to all TiKV. PD should scan keys to check that no non-TiDB data exists while switching `enable_api_v2`.
+
+### TiKV
+
+Add a new configuatrion `storage.enable_api_v2`. When enabled, `storage.enable_ttl` must also be enabled. TiKV will carry it's `storage.enable_api_v2` in `PutStore` request. PD will reject the `PutStore` if the switch mismatched.
```

Comment:
Can we just remove `storage.enable_ttl`?

Reply:
@zz-jason

Reply:
I think it's not appropriate to remove it but we can make it enable by default.